tools/codegen/api/python.py

```diff
@@ -0,0 +1,955 @@
+    def argument_str(self, *, method: bool = False) -> str:
+        name = self.name
+        typename = _simple_type(self.cpp_type_str)
+
+        # [old codegen] TODO: remove this and make optional types in simple_type
+        # to be consistent across tensor and other types after make Tensor? be
+        # optional instead of undefined
+        if self.type.is_nullable() and '?' not in typename:
+            typename = '{}?'.format(typename)
+
+        # s/self/input/ outside method bindings
+        # [old codegen] TODO: remove this? doesn't rename in codegen, it's just
+        # for the parse string
+        if name == 'self' and typename == 'Tensor' and not method:
+            name = 'input'
+
+        # add list size annotation
+        size = self.size
+        if size is not None:
+            if typename.endswith('?'):
+                typename = f'{typename[:-1]}[{size}]?'
+            else:
+                typename = f'{typename}[{size}]'
+
+        # add default
+        if self.default is not None:
+            default = {
+                'nullptr': 'None',
+                'c10::nullopt': 'None',
+                '{}': 'None',
+            }.get(self.default, self.default)
+            return f'{typename} {name}={default}'
+        else:
+            return f'{typename} {name}'
+
+    @property
+    def size(self) -> Optional[int]:
+        l = self.type.is_list_like()
+        return l.size \
+            if l is not None and l.size is not None and str(l.elem) != 'bool' else None
+
+
+@dataclass(frozen=True)
+class PythonOutArgument(PythonArgument):
+    # In Python signature multiple output fields are packed into one 'out' argument.
+    # When binding to C++, it's first binded to a local 'out' variable:
+    #   'auto out = _r.tensorlist_n<2>(2);',
+    # then binded to scattered C++ output arguments as 'out[0]', 'out[1]', and etc.
+    # TODO: maybe don't need keep scattered out fields for python signature?
+    outputs: Tuple[PythonArgument, ...]
+
+    @staticmethod
+    def from_outputs(outputs: Tuple[PythonArgument, ...]) -> Optional['PythonOutArgument']:
+        if not outputs:
+            return None
+
+        size = len(outputs)
+        if size == 1:
+            return PythonOutArgument(
+                name=outputs[0].name,
+                type=outputs[0].type,
+                cpp_type_str=outputs[0].cpp_type_str,
+                default='None',
+                default_init=None,
+                outputs=outputs,
+            )
+        elif size > 1:
+            return PythonOutArgument(
+                name='out',
+                type=ListType(BaseType(BaseTy.Tensor), size),
+                cpp_type_str='TensorList',
+                default='None',
+                default_init=None,
+                outputs=outputs,
+            )
+        raise AssertionError(r'Unexpected PythonOutArgument size')
+
+
+@dataclass(frozen=True)
+class PythonSignature:
```

Comment:
Here's a minor suggestion: because every PythonSignature derives from exactly one FunctionSchema, we should store the FunctionSchema in PythonSignature itself. This saves you from having to pass in a contextual FunctionSchema, as you do in bind_python_cpp. Curious what you think; @smessmer and I have argued about this point on other PRs and I'm not sure I've completely convinced him that keeping these around is a good idea.

Reply:
It's something I was thinking of as well - glad you also suggested it!
Because of deprecated.yaml, PythonSignature (instead of NativeFunction) has 1-1 mapping with Declarations.yaml entities. So it would be cleaner for the python code generator to all process the list of PythonSignatures and have NativeFunction included inside.

Reply:
> This saves you from having to pass in a contextual FunctionSchema, as you do in bind_python_cpp.

bind_python_cpp is a bit tricky though - we do bind two native functions (base version and out-variant) with one python signature, e.g.:

```
  static PythonArgParser parser({
    "eq(Tensor input, Tensor other, *, Tensor out=None)",
    ...
  });

...
  ParsedArgs<3> parsed_args;
  auto _r = parser.parse(nullptr, args, kwargs, parsed_args);
...

      if (_r.isNone(2)) {
        // aten::eq.Tensor(Tensor self, Tensor other) -> Tensor
        auto dispatch_eq = [](const Tensor & self, const Tensor & other) -> Tensor {
          pybind11::gil_scoped_release no_gil;
          return self.eq(other);
        };
        return wrap(dispatch_eq(_r.tensor(0), _r.tensor(1)));
      } else {
        // aten::eq.Tensor_out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
        auto dispatch_eq_out = [](Tensor out, const Tensor & self, const Tensor & other) -> Tensor {
          pybind11::gil_scoped_release no_gil;
          return at::eq_out(out, self, other);
        };
        return wrap(dispatch_eq_out(_r.tensor(2), _r.tensor(0), _r.tensor(1)));
      }
```

In the old codegen, we used two Declaration entities to represent the base case and the out variant, and used the `output_gap` to compensate the index difference between these two versions - because they are indeed consuming the output of the same python parser.
Here I plan to explicitly pass in the same PythonSignature and different NativeFunction variants to make it more explicit, e.g.:
```
bind(common_python_sig, base_native_function);
bind(common_python_sig, out_native_function);
```

Reply:
> What about holding a NativeFunction directly and just making separate PythonSignature instances for the two cases?

hmm, shouldn't we hold a PythonSignature and making separate NativeFunction instances for the two cases?

or did you mean we should pass in two NativeFunction instances and put (the same) PythonSignature inside both of them (to avoid passing PythonSignature separately)? That's probably fine - just that two PythonSignature instances might map to the same NativeFunction - one deprecated / one not. So PythonSignature is having 1-1 mapping with Declaration.yaml entities (which are what the python codegen processes).

So PythonSignature and NativeFunction do seem to have n-m mapping, no one is strictly belonging to the other.